influx-reports/influx.py

```diff
@@ -0,0 +1,140 @@
+    python time object. Shift it from
+    a few seconds (delta_seconds) if necessary.
+    """
+    timestamp_query = client.query(query)
+    # Get points is generator, we should just get first
+    # point time (string type).
+    fpt_str = next(timestamp_query.get_points())['time']
+    try:
+        fpt = dateutil.parser.parse(fpt_str) + datetime.timedelta(seconds=delta_seconds)
+    except Exception as e:
+        raise SystemExit(e)
+    return fpt
+
+
+def return_uptime(client, table, job_reference, resolution=60):
+    """ For a certain build (job_reference), estimate the amount
+    of seconds a component (column of table) was up.
+    For that, we slice the job by resolution.
+    If I have no data in the slice, I assume downtime. If I have
+    at least one data per slice, I assume up, even if degraded.
+    The count of these positive events multiplied by resolution
+    is an average idea of the uptime.
+    """
+    uptimes = defaultdict(int)
+
+    query = ("select max(/.*_status/) from {} where time < now() and "
+             "job_reference='{}' group by time({}s) fill(-1)"
+            ).format(table,job_reference,resolution)
+    all_data = client.query(query)
+    for time_slice in all_data.get_points():
+       for element in time_slice:
+           if element != 'time' and time_slice[element] == 1:
+               uptimes[element.replace('max_','')] += (1 * resolution)
+    return uptimes
+
+def main(args=None):
+    try:
+      influx_ip = os.environ['INFLUX_IP']
+    except KeyError:
+      logging.error("Please set INFLUX_IP")
+      sys.exit(1)
+
+    try:
+      influx_port = os.environ['INFLUX_PORT']
+    except KeyError:
+      logging.error("Please set INFLUX_PORT")
+      sys.exit(2)
+
+    try:
+      job_reference = os.environ['BUILD_TAG']
+    except KeyError:
+      logging.error("Please set BUILD_TAG for its usage as job ref")
+      sys.exit(3)
+
+    client = InfluxDBClient(influx_ip, influx_port, database='telegraf')
+
+    # First find the first and last timestamp from telegraf. This way querying maas_* data will always
+    # be accurate (if no data is reported by maas plugins, this should be a failure metric)
+    find_first_timestamp_query = ("select first(total) from processes "
+                                  "where job_reference = '{}';".format(job_reference))
+    first_ts = return_time(client, find_first_timestamp_query, delta_seconds=5)
+    find_last_timestamp_query = ("select last(total) from processes "
+                                 "where job_reference = '{}';".format(job_reference))
+    last_ts = return_time(client, find_last_timestamp_query, delta_seconds=-5)
+
+    logging.info("Metrics were gathered between {} and {}".format(first_ts, last_ts))
+
+    measurements = ['maas_glance','maas_cinder','maas_keystone', 'maas_heat', 'maas_neutron',
+                    'maas_nova', 'maas_horizon']
+
+    yml_report = False
+    subunit_report = False
+
+    if args.ymlreport:
```

Comment:
I'd wrap this whole section in try/finally to ensure `output.stopTestRun` gets called. You could use context managers and exitstack, seeming as you're in py3.  https://docs.python.org/3/library/contextlib.html#contextlib.ExitStack